crates/uv-requirements/src/lookahead.rs

```diff
@@ -0,0 +1,104 @@
+use std::sync::Arc;
+
+use anyhow::{Context, Result};
+use futures::{StreamExt, TryStreamExt};
+
+use distribution_types::{BuildableSource, Dist, RequestedRequirements};
+use pep508_rs::{Requirement, VersionOrUrl};
+use uv_client::RegistryClient;
+use uv_distribution::{Reporter, SourceDistCachedBuilder};
+use uv_traits::BuildContext;
+
+/// A resolver for resolving lookahead requirements from local dependencies.
+///
+/// The resolver extends certain privileges to "first-party" requirements. For example, first-party
+/// requirements are allowed to contain direct URL references, local version specifiers, and more.
+///
+/// We make an exception for transitive requirements of _local_ dependencies. For example,
+/// `pip install .` should treat the dependencies of `.` as if they were first-party dependencies.
+/// This matches our treatment of editable installs (`pip install -e .`).
+///
+/// The lookahead resolver resolves requirements for local dependencies, so that the resolver can
+/// treat them as first-party dependencies for the purpose of analyzing their specifiers.
+pub struct LookaheadResolver<'a> {
```

Comment:
The alternative solution would be to pass any source trees to `SourceTreeResolver` (so, e.g., if the user does `foo @ ./bar/baz`, we'd pass `./bar/baz` to `SourceTreeResolver` alongside any `pyproject.toml` or `setup.py` files that were passed in). Then, we'd treat the requirements as part of the `requirements` input to the resolution.

That approach would require some refactoring of `SourceTreeResolver` (since it uses `ExtrasSpecification`, which isn't a generic concept for requirements). It would also mean that the actual `Resolver` wouldn't be able to distinguish between these "lookahead" requirements and actual first-party requirements, and there may not be any indication in the error messages that they came from the local directory requirement, but that might be ok. I'm open to either.

Reply:
\cc @zanieb

Reply:
This seems okay, but the whole `SourceTree` and `SourceTreeResolver` concept is so new that I don't have strong feelings.

I like the idea of them being included in the resolver manifest.